### ISSUE TITLE
third field vs. fourth field

### ISSUE THREAD

**Issue by kosoegawa** [nmdp-bioinformatics/ImmunogeneticDataTools#5]
If we can find perfect match including fourth field, it is great. If Linkage not found, I would like to trim the fourth field and compare three field level. It is good to have a report which field comparison is made for the results.

**Comment by kosoegawa** [comment]
There are two potential issues.
1. table has fourth field, and our data has fourth filed, but the difference is in only fourth filed.
2. Table data may have only three field, but GL string may have fourth filed.
Or Table data may have fourth filed, but GL string may have only three field.

**Comment by mpresteg** [comment]
Report on various levels of linkage found (3rd or 4th field).

**Comment by kosoegawa** [comment]
Feb 20, 2015 4:54:11 PM org.dash.valid.LinkageDisequilibriumWriter reportDetectedLinkages
INFO: Your GL String: HLA-A_02:01:01:01+HLA-A_03:01:01:01/HLA-A_03:01:01:02N/HLA-A_03:01:01:03^HLA-B_14:02:01+HLA-B_52:01:01:01/HLA-B_52:01:01:02^HLA-C_08:02:01/HLA-C_08:34+HLA-C_12:02:02^HLA-DPA1_01:03:01:01/HLA-DPA1_01:03:01:02/HLA-DPA1_01:03:01:03/HLA-DPA1_01:03:01:04/HLA-DPA1_01:03:01:05+HLA-DPA1_01:03:01:01/HLA-DPA1_01:03:01:02/HLA-DPA1_01:03:01:03/HLA-DPA1_01:03:01:04/HLA-DPA1_01:03:01:05^HLA-DPB1_04:01:01:01/HLA-DPB1_04:01:01:02+HLA-DPB1_04:02:01:01/HLA-DPB1_04:02:01:02|HLA-DPB1_105:01+HLA-DPB1_126:01^HLA-DQA1_01:03:01:01/HLA-DQA1_01:03:01:02+HLA-DQA1_05:03/HLA-DQA1_05:07^HLA-DQB1_03:01:01:01/HLA-DQB1_03:01:01:02/HLA-DQB1_03:01:01:03+HLA-DQB1_06:01:01^HLA-DRB1_11:01:01+HLA-DRB1_15:02:01^HLA-DRB3_02:02:01:01/HLA-DRB3_02:02:01:02^HLA-DRB5*01:02

We found linkages:
DRB1 Locus: HLA-DRB1_15:02:01
DRB345 Locus: HLA-DRB5_01:02
DQA1 Locus: HLA-DQA1_01:03:01:01
DQB1 Locus: HLA-DQB1_06:01:01
NegLocFreq: 2.2
Notes:  

We found linkages:
B Locus: HLA-B_14:02:01
C Locus: HLA-C_08:02:01
NegLocFreq: 1.6
Notes:  

We found linkages:
B Locus: HLA-B_52:01:01
C Locus: HLA-C_12:02:02
NegLocFreq: 2.1
Notes:

**Comment by kosoegawa** [comment]
The above example misses
DRB1*11:01 linkage

**Comment by mpresteg** [comment]
I'm not seeing that linkage.  Could you point explicitly to the linkage row (what are the other values)?

**Comment by kosoegawa** [comment]
Feb 20, 2015 10:16:18 PM org.dash.valid.LinkageDisequilibriumWriter reportDetectedLinkages
INFO: Your GL String: HLA-A_02:01:01:01+HLA-A_24:02:01:01/HLA-A_24:02:01:03^HLA-B_27:02:01+HLA-B_35:02:01/HLA-B_35:211^HLA-C_02:02:02+HLA-C_04:01:01:01/HLA-C_04:01:01:02/HLA-C_04:01:01:03/HLA-C_04:01:01:04/HLA-C_04:01:01:05/HLA-C_04:20/HLA-C_04:117^HLA-DPA1_01:03:01:01/HLA-DPA1_01:03:01:02/HLA-DPA1_01:03:01:03/HLA-DPA1_01:03:01:04/HLA-DPA1_01:03:01:05+HLA-DPA1_02:01:01^HLA-DPB1_04:01:01:01/HLA-DPB1_04:01:01:02+HLA-DPB1_17:01^HLA-DQA1_02:01+HLA-DQA1_05:05:01:01/HLA-DQA1_05:05:01:02/HLA-DQA1_05:05:01:03/HLA-DQA1_05:09/HLA-DQA1_05:11^HLA-DQB1_02:02+HLA-DQB1_03:01:01:01/HLA-DQB1_03:01:01:02/HLA-DQB1_03:01:01:03^HLA-DRB1_07:01:01:01/HLA-DRB1_07:01:01:02+HLA-DRB1_11:01:01^HLA-DRB3_01:01:02:01/HLA-DRB3_01:01:02:02^HLA-DRB4_01:03:01:01/HLA-DRB4_01:03:01:03

We found partial linkages:
DRB1 Locus: HLA-DRB1_07:01:01:01
DRB345 Locus: HLA-DRB4_01:03:01:02N
DQA1 Locus: HLA-DQA1_02:01
DQB1 Locus: HLA-DQB1_02:02
NegLocFreq: 3.5
Notes:  

We found perfect linkages:
B Locus: HLA-B_35:02:01
C Locus: HLA-C_04:01:01:01
NegLocFreq: 2.1
Notes:  

We found perfect linkages:
DRB1 Locus: HLA-DRB1_07:01:01:01
DRB345 Locus: HLA-DRB4_01:03:01:01
DQA1 Locus: HLA-DQA1_02:01
DQB1 Locus: HLA-DQB1_02:02
NegLocFreq: 1.4
Notes:  

We found perfect linkages:
DRB1 Locus: HLA-DRB1_11:01:01
DRB345 Locus: HLA-DRB3_01:01:02:01
DQA1 Locus: HLA-DQA1_05:05:01:01
DQB1 Locus: HLA-DQB1_03:01:01:01
NegLocFreq: 3.2
Notes:

**Comment by kosoegawa** [comment]
The above example missed linkage as follows:
In the look up table (BCLinkageDisequilibrium.txt), it says:
HLA-B_27:02      HLA-C_02:02:02  2.6
The program could not find this.
My typing example says:
HLA-B_27:02:01 HLA-C_02:02:02

**Comment by kosoegawa** [comment]
When I submitted separator (*) disappeared. I hope you can figure it out.

**Comment by kosoegawa** [comment]
I talked to Steve Mack on my way back home yesterday. He is excited about this program, but he would like us to use more updated lookup table. He is a population geneticist. I am hoping it is not so difficult to replace the table as long as we keep the same format. I will get the table from Steve. Also he suggested that we might want to include DPA1/DPB1 linkage as well. I do not have an example, so I will get it from Steve.

**Comment by kosoegawa** [comment]
It would be good if "Linkage Not Found" could be included in case we could not find linkage like above example.

**Comment by kosoegawa** [comment]
Feb 20, 2015 10:16:18 PM org.dash.valid.LinkageDisequilibriumWriter reportDetectedLinkages
INFO: Your GL String: HLA-A_02:01:01:01+HLA-A_03:01:01:01/HLA-A_03:01:01:02N/HLA-A_03:01:01:03^HLA-B_14:02:01+HLA-B_52:01:01:01/HLA-B_52:01:01:02^HLA-C_08:02:01/HLA-C_08:34+HLA-C_12:02:02^HLA-DPA1_01:03:01:01/HLA-DPA1_01:03:01:02/HLA-DPA1_01:03:01:03/HLA-DPA1_01:03:01:04/HLA-DPA1_01:03:01:05+HLA-DPA1_01:03:01:01/HLA-DPA1_01:03:01:02/HLA-DPA1_01:03:01:03/HLA-DPA1_01:03:01:04/HLA-DPA1_01:03:01:05^HLA-DPB1_04:01:01:01/HLA-DPB1_04:01:01:02+HLA-DPB1_04:02:01:01/HLA-DPB1_04:02:01:02|HLA-DPB1_105:01+HLA-DPB1_126:01^HLA-DQA1_01:03:01:01/HLA-DQA1_01:03:01:02+HLA-DQA1_05:03/HLA-DQA1_05:07^HLA-DQB1_03:01:01:01/HLA-DQB1_03:01:01:02/HLA-DQB1_03:01:01:03+HLA-DQB1_06:01:01^HLA-DRB1_11:01:01+HLA-DRB1_15:02:01^HLA-DRB3_02:02:01:01/HLA-DRB3_02:02:01:02^HLA-DRB5*01:02

We found partial linkages:
B Locus: HLA-B_52:01:01
C Locus: HLA-C_12:02:02
NegLocFreq: 2.1
Notes:  

We found perfect linkages:
DRB1 Locus: HLA-DRB1_15:02:01
DRB345 Locus: HLA-DRB5_01:02
DQA1 Locus: HLA-DQA1_01:03:01:01
DQB1 Locus: HLA-DQB1_06:01:01
NegLocFreq: 2.2
Notes:  

We found perfect linkages:
B Locus: HLA-B_14:02:01
C Locus: HLA-C_08:02:01
NegLocFreq: 1.6
Notes:

**Comment by kosoegawa** [comment]
The software did good job in terms "Linkage Not Found".
In my example:
HLA-DRB1_11:01:01
HLA-DRB3_02:02:01:01
HLA-DQA1_05:03
HLA-DQB1_03:01:01:01

I could not find DQA1_05:03 in the lookup table. I found DQA1_05:05:01:01.
As I said above, it is nice if we could see "WARNING - Linkage Not Found", since this could be unusual haplotype, and we may have typing error. This is exactly what we are looking for.

**Comment by mpresteg** [comment]
Closing this and creating new discrete issue.

RESOLUTION:
- Original issue described conditions related to the presence or absence of the 4th field.  Subsequent examples imply similar logic needed for the presence or absence of the 3rd field (when only 2nd field is present).  Modified alleleShortening method to instead do field level comparison dependent upon how many positions are populated in either the allele or the reference allele (2, 3 or 4).

New output for provided example:

We found perfect linkages:
B Locus: HLA-B*35:02:01
C Locus: HLA-C*04:01:01:01
NegLocFreq: 2.1
Notes:  

We found partial linkages:
B Locus: HLA-B*27:02
C Locus: HLA-C*02:02:02
NegLocFreq: 2.6
Notes:  

We found perfect linkages:
DRB1 Locus: HLA-DRB1*07:01:01:01
DRB345 Locus: HLA-DRB4*01:03:01:01
DQA1 Locus: HLA-DQA1*02:01
DQB1 Locus: HLA-DQB1*02:02
NegLocFreq: 1.4
Notes:  

We found perfect linkages:
DRB1 Locus: HLA-DRB1*11:01:01
DRB345 Locus: HLA-DRB3*01:01:02:01
DQA1 Locus: HLA-DQA1*05:05:01:01
DQB1 Locus: HLA-DQB1*03:01:01:01
NegLocFreq: 3.2
Notes:  

We found partial linkages:
DRB1 Locus: HLA-DRB1*07:01:01:01
DRB345 Locus: HLA-DRB4*01:03:01:02N
DQA1 Locus: HLA-DQA1*02:01
DQB1 Locus: HLA-DQB1*02:02
NegLocFreq: 3.5
Notes: 

NEW ISSUES:
- Request to display "WARNING - Linkage Not Found".
- Request for new/updated reference data set.

COMMENT:
For future reference - In order to avoid the * being stripped off they must be escaped with \
